packages/ts/react-signals/src/Signals.ts

```diff
@@ -0,0 +1,78 @@
+import { Signal } from './core.js';
+import type { SetEvent, StateEvent } from './types';
+
+/**
+ * A signal that holds a value. The underlying
+ * value of this signal is stored and updated as a
+ * shared value on the server.
+ *
+ * @internal
+ */
+export abstract class ValueSignal<T> extends Signal<T> {
+  private readonly publish: (event: StateEvent) => Promise<boolean>;
+
+  /**
+   * Creates a new ValueSignal instance.
+   * @param publish The function that publishes the
+   * value of the signal to the server.
+   * @param value The initial value of the signal
+   * @defaultValue undefined
+   */
+  constructor(publish: (event: StateEvent) => Promise<boolean>, value?: T) {
+    super(value);
+    this.publish = publish;
+  }
+
+  /**
+   * Returns the value of the signal.
+   */
+  override get value() {
+    return super.value;
+  }
+
+  /**
+   * Publishes the new value to the server.
+   * Note that this method is not setting
+   * the signal's value.
+   *
+   * @param value The new value of the signal
+   * to be published to the server.
+   */
+  override set value(value: T) {
+    const id = crypto.randomUUID();
+    const event: SetEvent = { id, type: 'set', value };
+    this.publish(event).then(r => undefined);
```

Comment:
I'd suggest merging it together as it doesn't bring any value to split them up.
```suggestion
    this.publish({ id, type: 'set', value }).then(r => undefined);
```